hal/src/main/native/athena/SerialPort.cpp

```diff
@@ -1,174 +1,461 @@
 }
 
-void HAL_SetSerialTimeout(HAL_SerialPort port, double timeout,
+void HAL_SetSerialTimeout(HAL_SerialPortHandle handle, double timeout,
                           int32_t* status) {
-  *status = viSetAttribute(portHandles[port], VI_ATTR_TMO_VALUE,
-                           static_cast<uint32_t>(timeout * 1e3));
-  if (*status > 0) *status = 0;
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return;
+  }
+  port->timeout = timeout;
+  port->tty.c_cc[VTIME] = static_cast<int>(timeout * 10);
+  tcsetattr(port->portId, TCSANOW, &port->tty);
 }
 
-void HAL_EnableSerialTermination(HAL_SerialPort port, char terminator,
+void HAL_EnableSerialTermination(HAL_SerialPortHandle handle, char terminator,
                                  int32_t* status) {
-  viSetAttribute(portHandles[port], VI_ATTR_TERMCHAR_EN, VI_TRUE);
-  viSetAttribute(portHandles[port], VI_ATTR_TERMCHAR, terminator);
-  *status = viSetAttribute(portHandles[port], VI_ATTR_ASRL_END_IN,
-                           VI_ASRL_END_TERMCHAR);
-  if (*status > 0) *status = 0;
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return;
+  }
+  port->termination = true;
+  port->terminationChar = terminator;
 }
 
-void HAL_DisableSerialTermination(HAL_SerialPort port, int32_t* status) {
-  viSetAttribute(portHandles[port], VI_ATTR_TERMCHAR_EN, VI_FALSE);
-  *status =
-      viSetAttribute(portHandles[port], VI_ATTR_ASRL_END_IN, VI_ASRL_END_NONE);
-  if (*status > 0) *status = 0;
+void HAL_DisableSerialTermination(HAL_SerialPortHandle handle,
+                                  int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return;
+  }
+  port->termination = false;
 }
 
-void HAL_SetSerialReadBufferSize(HAL_SerialPort port, int32_t size,
+void HAL_SetSerialReadBufferSize(HAL_SerialPortHandle handle, int32_t size,
                                  int32_t* status) {
-  *status = viSetBuf(portHandles[port], VI_READ_BUF, size);
-  if (*status > 0) *status = 0;
+  // NO OP currently
 }
 
-void HAL_SetSerialWriteBufferSize(HAL_SerialPort port, int32_t size,
+void HAL_SetSerialWriteBufferSize(HAL_SerialPortHandle handle, int32_t size,
                                   int32_t* status) {
-  *status = viSetBuf(portHandles[port], VI_WRITE_BUF, size);
-  if (*status > 0) *status = 0;
+  // NO OP currently
 }
 
-int32_t HAL_GetSerialBytesReceived(HAL_SerialPort port, int32_t* status) {
-  int32_t bytes = 0;
-
-  *status = viGetAttribute(portHandles[port], VI_ATTR_ASRL_AVAIL_NUM, &bytes);
-  if (*status > 0) *status = 0;
+int32_t HAL_GetSerialBytesReceived(HAL_SerialPortHandle handle,
+                                   int32_t* status) {
+  auto port = serialPortHandles->Get(handle);
+  if (!port) {
+    *status = HAL_HANDLE_ERROR;
+    return -1;
+  }
+  int bytes;
+  ioctl(port->portId, FIONREAD, &bytes);
```

Comment:
If this ioctl fails, bytes will be left uninitialized.  Should check ioctl return value and/or init bytes to 0.